### PR TITLE
Adds a `multiply_grads` akin to fairseq

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Helps w/ DDP wrt https://github.com/huggingface/transformers/pull/34283

`fairseq` came up with a technique to help also with countering gradient accumulation by countering grad update by multiplying the gradients by `num_gpus / num_items_in_batch`. This PR adds an interface for the `Trainer` to hook into. 

Without it:

W B Chart 10_22_2024, 12_30_21 PM

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@SunMarc

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
Nice ! It would be also nice to compare if it is better to update the optimizer or the loss directly in transformers.

**Comment by sbwww** [comment]
There is a minor concern, would this fix be suitable for all backends (PyTorch DDP/FDSP, DeepSpeed, Megatron)?

I notice that DeepSpeed optimizer has unique groups (`bf16_groups` and `fp32_groups_gradients`) in its wrapped optimizer. And, grads in `bf16_groups` are accumulated into `fp32_groups_gradients` and deallocated in each `backward` call, .

I'm not sure whether `param_groups` in accelerator is a wrap of `bf16_groups` or `fp32_groups_gradients` in DeepSpeed? Previously I had an attempt to manipulate grads in optimizer. However, grads in `optimizer` are all 0 after backward when I use accelerate + DeepSpeed.

So, I'm afraid `multiply_grads` would not work for current DeepSpeed integration.

---

DeepSpeed/deepspeed/runtime/bf16_optimizer.py :

https://github.com/microsoft/DeepSpeed/blob/b647fb2470f8f6fefe5cab0ea84a2d89696eb898/deepspeed/runtime/bf16_optimizer.py#L316-L329
